usd/src/usd_parser/USDTransforms.cc

```diff
@@ -0,0 +1,395 @@
+}
+
+//////////////////////////////////////////////////
+bool UDSTransforms::Rotation() const
+{
+  return this->dataPtr->isRotation;
+}
+
+//////////////////////////////////////////////////
+void UDSTransforms::RotationZYX(bool _rotationZYX)
+{
+  this->dataPtr->isRotationZYX = _rotationZYX;
+}
+
+//////////////////////////////////////////////////
+void UDSTransforms::RotationXYZ(bool _rotationXYZ)
+{
+  this->dataPtr->isRotationXYZ = _rotationXYZ;
+}
+
+//////////////////////////////////////////////////
+void UDSTransforms::Rotation(bool _rotation)
+{
+  this->dataPtr->isRotation = _rotation;
+}
+
+//////////////////////////////////////////////////
+void GetAllTransforms(
+  const pxr::UsdPrim &_prim,
+  const USDData &_usdData,
+  std::vector<ignition::math::Pose3d> &_tfs,
+  ignition::math::Vector3d &_scale,
+  const std::string &_schemaToStop)
+{
+  pxr::UsdPrim parent = _prim;
+  double metersPerUnit = 1.0;
+  std::string upAxis = "Y";
+
+  // this assumes that there can only be one stage
+  const auto stageData = _usdData.FindStage(parent.GetPath().GetName());
+  if (stageData.second != nullptr) {
+    metersPerUnit = stageData.second->MetersPerUnit();
+    upAxis = stageData.second->UpAxis();
+  }
+
+  while (parent)
+  {
+    if (pxr::TfStringify(parent.GetPath()) == _schemaToStop)
+    {
+      return;
+    }
+
+    UDSTransforms t = ParseUSDTransform(parent);
+
+    ignition::math::Pose3d pose;
+    _scale *= t.Scale();
+
+    pose.Pos() = t.Translate() * metersPerUnit;
+    // scaling is lost when we convert to pose, so we pre-scale the
+    // translation to make them match the scaled values.
+    if (!_tfs.empty()) {
+      auto& child = _tfs.back();
+      child.Pos().Set(
+        child.Pos().X() * t.Scale()[0],
+        child.Pos().Y() * t.Scale()[1],
+        child.Pos().Z() * t.Scale()[2]);
+    }
+
+    if (!t.RotationZYX() && !t.RotationXYZ())
+    {
+      if (t.Rotation())
+      {
+        pose.Rot() = t.Rotations()[0];
+      }
+      _tfs.push_back(pose);
+    }
+    else
+    {
+      ignition::math::Pose3d poseZ = ignition::math::Pose3d(
+        ignition::math::Vector3d(0, 0, 0), t.Rotations()[2]);
+      ignition::math::Pose3d poseY = ignition::math::Pose3d(
+        ignition::math::Vector3d(0, 0, 0), t.Rotations()[1]);
+      ignition::math::Pose3d poseX = ignition::math::Pose3d(
+        ignition::math::Vector3d(0, 0, 0), t.Rotations()[0]);
+
+      ignition::math::Pose3d poseT = ignition::math::Pose3d(
+        t.Translate() * metersPerUnit,
+        ignition::math::Quaterniond(1, 0, 0, 0));
+
+      _tfs.push_back(poseZ);
+      _tfs.push_back(poseY);
+      _tfs.push_back(poseX);
+      _tfs.push_back(poseT);
+    }
+    parent = parent.GetParent();
+  }
+
+  if (upAxis == "Y")
+  {
+    ignition::math::Pose3d poseUpAxis = ignition::math::Pose3d(
+      ignition::math::Vector3d(0, 0, 0),
+      ignition::math::Quaterniond(IGN_PI_2, 0, 0));
+    _tfs.push_back(poseUpAxis);
+  }
```

Comment:
How was this resolved?

Reply:
I added this comment 

`// Add additional rotation to match with Z up Axis`

Reply:
afaik, this transform is to convert the up axis from Y to Z. If `upAxis` is already Z there is no need to add the extra transform. Technically there could be a `upAxis = X` but I don't think any "normal" renderer uses that.

Reply:
Okay, thanks for the explanation.

Reply:
I left a `TODO` note about handling `upAxis == "X"` in https://github.com/ignitionrobotics/sdformat/pull/871/commits/e2acde3f93cc5be5bb136851cc8f813e6753d430 in case we decide to revisit this later, but it sounds like handling `Y` and `Z` are sufficient for now. I will mark this as resolved.